src/Avalonia.X11/Clipboard/X11Clipboard.cs

```diff
@@ -136,105 +140,78 @@ IntPtr WriteTargetToProperty(IntPtr target, IntPtr window, IntPtr property)
-                            Marshal.Copy(prop, data, 0, data.Length);
-                            _requestedDataTcs?.TrySetResult(data);
-                        }
-                    }
-                }
-
-                XFree(prop);
-            }
         }
 
-        private Task<IntPtr[]?> SendFormatRequest()
+        async void SendIncrDataToClient(IntPtr window, IntPtr property, IntPtr target, Stream data)
         {
-            if (_requestedFormatsTcs == null || _requestedFormatsTcs.Task.IsCompleted)
-                _requestedFormatsTcs = new TaskCompletionSource<IntPtr[]?>();
-            XConvertSelection(_x11.Display, _x11.Atoms.CLIPBOARD, _x11.Atoms.TARGETS, _x11.Atoms.TARGETS, _handle,
-                IntPtr.Zero);
-            return _requestedFormatsTcs.Task;
+            data.Position = 0;
+            using var events = new EventStreamWindow(_platform, window);
+            using var _ = data;
+            XSelectInput(_x11.Display, window, new IntPtr((int)XEventMask.PropertyChangeMask));
+            var size = new IntPtr(data.Length);
+            XChangeProperty(_x11.Display, window, property, _x11.Atoms.INCR, 32, PropertyMode.Replace, ref size, 1);
+            var buffer = ArrayPool<byte>.Shared.Rent((int)Math.Min(_maximumPropertySize, data.Length));
+            while (true)
+            {
+                if (null == await events.WaitForEventAsync(x =>
+                        x.type == XEventName.PropertyNotify && x.PropertyEvent.atom == property
+                                                            && x.PropertyEvent.state == 1, TimeSpan.FromMinutes(1)))
+                    break;
+                var read = await data.ReadAsync(buffer, 0, buffer.Length);
+                if(read == 0)
+                    break;
+                XChangeProperty(_x11.Display, window, property, target, 8, PropertyMode.Replace, buffer, read);
+            }
+            ArrayPool<byte>.Shared.Return(buffer);
+
+            // Finish the transfer
+            XChangeProperty(_x11.Display, window, property, target, 8, PropertyMode.Replace, IntPtr.Zero, 0);
         }
 
-        private Task<object?> SendDataRequest(IntPtr format)
+        void SendDataToClient(IntPtr window, IntPtr property, IntPtr target, byte[] bytes)
         {
-            if (_requestedDataTcs == null || _requestedDataTcs.Task.IsCompleted)
-                _requestedDataTcs = new TaskCompletionSource<object?>();
-            XConvertSelection(_x11.Display, _x11.Atoms.CLIPBOARD, format, format, _handle, IntPtr.Zero);
-            return _requestedDataTcs.Task;
+            if (bytes.Length < _maximumPropertySize)
+            {
+                XChangeProperty(_x11.Display, window, property, target, 8,
+                    PropertyMode.Replace,
+                    bytes, bytes.Length);
+            }
+            SendIncrDataToClient(window, property, target, new MemoryStream(bytes));
```

Comment:
Is an `else` missing? This will currently set the property twice (once normally, once with INCR) if the data doesn't need INCR.

Reply:
Looks like a missing else, yes